torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -168,15 +169,70 @@ void RequestCallbackImpl::processRpc(
     case MessageType::PYTHON_CALL: {
       auto& upc = static_cast<UnpickledPythonCall&>(rpc);
       auto& pythonRpcHandler = PythonRpcHandler::getInstance();
-      std::shared_ptr<SerializedPyObj> serializedPyObj = nullptr;
-      {
-        py::gil_scoped_acquire acquire;
-        serializedPyObj =
-            std::make_shared<SerializedPyObj>(pythonRpcHandler.serialize(
-                pythonRpcHandler.runPythonUdf(std::move(upc).movePythonUdf())));
+      if (upc.isAsyncFunction()) {
+        std::shared_ptr<jit::PythonFutureWrapper> pyFuture;
+        {
+          py::gil_scoped_acquire acquire;
+          auto result =
+              pythonRpcHandler.runPythonUdf(std::move(upc).movePythonUdf());
+
+          if (pythonRpcHandler.isException(result) ) {
+            // Hit exception when running the user function.
+            // Not releasing GIL before serialize to avoid an additional
+            // context switch.
+            auto serializedPyObj = pythonRpcHandler.serialize(result);
+            py::gil_scoped_release release;
+            markComplete(
+                std::move(PythonResp(std::move(serializedPyObj))).toMessage());
+            return;
+          }
+
+          try {
+            pyFuture = result.cast<std::shared_ptr<jit::PythonFutureWrapper>>();
+          } catch (...) {
+            // user function return type cannot be cast to a Future type.
+            auto type = result.get_type();
+            auto errMsg = c10::str(
+                "Functions decorated with @rpc.async_function must return a "
+                "torch.futures.Future object, but got ",
+                type.attr("__module__").cast<std::string>(),
+                ".",
+                type.attr("__qualname__").cast<std::string>());
+
+            py::gil_scoped_release release;
+            responseFuture->markCompleted(
+                createExceptionResponse(errMsg, messageId));
+            return;
+          }
+        }
+        pyFuture->fut->addCallback([msgId = messageId,
```

Comment:
So, with this change, functions decorated with `@rpc.rpc_async` will no longer block a thread on the server, correct? Do we need to update documentation anywhere to recommend users to use this in case their UDFs involve RPC or the like?

Reply:
Good point! Will add the doc change to `rpc.rst` when the PR for `rpc.remote` is also ready. If we add the change to API docs now, it will make `rpc_async`, `rpc_sync`, and `remote` inconsistent.

Reply:
Created #39399 to track